{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "63382fb9-15ee-460d-88c7-7ae9fb5ca747",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'modin'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "File \u001b[0;34m/home/milo/.cache/pypoetry/virtualenvs/isotools-kPChPiwz-py3.8/lib/python3.8/site-packages/IPython/core/interactiveshell.py\u001b[0m, in \u001b[0;32mrun_code\u001b[0m:\nLine \u001b[0;34m3441\u001b[0m:  exec(code_obj, \u001b[36mself\u001b[39;49;00m.user_global_ns, \u001b[36mself\u001b[39;49;00m.user_ns)\n",
      "In  \u001b[0;34m[1]\u001b[0m:\nLine \u001b[0;34m2\u001b[0m:     \u001b[34mimport\u001b[39;49;00m \u001b[04m\u001b[36mmodin\u001b[39;49;00m\u001b[04m\u001b[36m.\u001b[39;49;00m\u001b[04m\u001b[36mpandas\u001b[39;49;00m \u001b[34mas\u001b[39;49;00m \u001b[04m\u001b[36mpd\u001b[39;49;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'modin'\n\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"
     ]
    }
   ],
   "source": [
    "#import pandas as pd\n",
    "import modin.pandas as pd\n",
    "import gtfparse\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bddeea6f-0d1f-44f8-8884-0d8cd4dc28dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "chm13t2t = Path(\"/mnt/group/references/genomic/homo_sapiens/sequences/chm13t2t/chm13.draft_v1.1.gene_annotation.v4.gff3.gz\")\n",
    "gencode = Path(\"/mnt/group/references/genomic/homo_sapiens/sequences/gencode_v38/gencode.v38.chr_patch_hapl_scaff.basic.annotation.gff3.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a48359ac-4423-4d64-baf7-d714c7767225",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[parse_gtf(95)]: Reading in data in chunks\n",
      "loading file:  11%|#1        | 1/9 [06:15<50:01, 375.14s/chunks]"
     ]
    }
   ],
   "source": [
    "chm13t2t_anno = gtfparse.read_gtf(chm13t2t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fbbcec14-b8c0-4cd2-acf5-b991c4557f41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gffpandas.gffpandas.Gff3DataFrame at 0x7fa83f5eaf40>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13t2t_anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "85e8e9c5-9847-4205-ae88-11785ca5b5c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>gene</td>\n",
       "      <td>11136</td>\n",
       "      <td>12457</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name=AL627309.3;source_gene=ENSG000002399...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11136</td>\n",
       "      <td>12457</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name=AL627309.3;source_gene=ENSG000002399...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>exon</td>\n",
       "      <td>11136</td>\n",
       "      <td>11635</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name=AL627309.3;source_gene=ENSG000002399...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>gene</td>\n",
       "      <td>11630</td>\n",
       "      <td>13433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name=AP006222.2;source_gene=ENSG000002864...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11630</td>\n",
       "      <td>13433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name=AP006222.2;source_gene=ENSG000002864...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seq_id   source        type  start    end score strand phase  \\\n",
       "0   chr1  Liftoff        gene  11136  12457     .      -     .   \n",
       "1   chr1  Liftoff  transcript  11136  12457     .      -     .   \n",
       "2   chr1  Liftoff        exon  11136  11635     .      -     .   \n",
       "3   chr1  Liftoff        gene  11630  13433     .      +     .   \n",
       "4   chr1  Liftoff  transcript  11630  13433     .      +     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  gene_name=AL627309.3;source_gene=ENSG000002399...  \n",
       "1  gene_name=AL627309.3;source_gene=ENSG000002399...  \n",
       "2  gene_name=AL627309.3;source_gene=ENSG000002399...  \n",
       "3  gene_name=AP006222.2;source_gene=ENSG000002864...  \n",
       "4  gene_name=AP006222.2;source_gene=ENSG000002864...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13t2t_anno.df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "849a9d76-9290-41e5-804d-5250e0f2d773",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_anno = gffpd.read_gff3(gencode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2ac1cd43-085c-4adf-9d3c-ab14576effce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000456328.2;Parent=ENSG00000223972.5;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:1;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:2;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000456328.2:3;Parent=ENST0000045...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043934</th>\n",
       "      <td>KI270744.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>51009</td>\n",
       "      <td>51114</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000616830.1;Parent=ENSG00000278625.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043935</th>\n",
       "      <td>KI270744.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>51009</td>\n",
       "      <td>51114</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000616830.1:1;Parent=ENST0000061...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043936</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENSG00000277374.1;gene_id=ENSG00000277374.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043937</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=ENST00000612925.1;Parent=ENSG00000277374.1;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043938</th>\n",
       "      <td>KI270750.1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>148668</td>\n",
       "      <td>148843</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=exon:ENST00000612925.1:1;Parent=ENST0000061...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2043939 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             seq_id   source        type   start     end score strand phase  \\\n",
       "0              chr1   HAVANA        gene   11869   14409     .      +     .   \n",
       "1              chr1   HAVANA  transcript   11869   14409     .      +     .   \n",
       "2              chr1   HAVANA        exon   11869   12227     .      +     .   \n",
       "3              chr1   HAVANA        exon   12613   12721     .      +     .   \n",
       "4              chr1   HAVANA        exon   13221   14409     .      +     .   \n",
       "...             ...      ...         ...     ...     ...   ...    ...   ...   \n",
       "2043934  KI270744.1  ENSEMBL  transcript   51009   51114     .      -     .   \n",
       "2043935  KI270744.1  ENSEMBL        exon   51009   51114     .      -     .   \n",
       "2043936  KI270750.1  ENSEMBL        gene  148668  148843     .      +     .   \n",
       "2043937  KI270750.1  ENSEMBL  transcript  148668  148843     .      +     .   \n",
       "2043938  KI270750.1  ENSEMBL        exon  148668  148843     .      +     .   \n",
       "\n",
       "                                                attributes  \n",
       "0        ID=ENSG00000223972.5;gene_id=ENSG00000223972.5...  \n",
       "1        ID=ENST00000456328.2;Parent=ENSG00000223972.5;...  \n",
       "2        ID=exon:ENST00000456328.2:1;Parent=ENST0000045...  \n",
       "3        ID=exon:ENST00000456328.2:2;Parent=ENST0000045...  \n",
       "4        ID=exon:ENST00000456328.2:3;Parent=ENST0000045...  \n",
       "...                                                    ...  \n",
       "2043934  ID=ENST00000616830.1;Parent=ENSG00000278625.1;...  \n",
       "2043935  ID=exon:ENST00000616830.1:1;Parent=ENST0000061...  \n",
       "2043936  ID=ENSG00000277374.1;gene_id=ENSG00000277374.1...  \n",
       "2043937  ID=ENST00000612925.1;Parent=ENSG00000277374.1;...  \n",
       "2043938  ID=exon:ENST00000612925.1:1;Parent=ENST0000061...  \n",
       "\n",
       "[2043939 rows x 9 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_anno.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4db7ba92-4762-45d9-bbb5-98277d44545f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_df = pd.read_csv(gencode, delimiter=\"\\t\", nrows=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "77402fdb-0b48-436b-9139-60f11abd0c95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>##gff-version 3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">chr1</th>\n",
       "      <th rowspan=\"5\" valign=\"top\">Liftoff</th>\n",
       "      <th>gene</th>\n",
       "      <th>11136</th>\n",
       "      <th>12457</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_name=AL627309.3;source_gene=ENSG000002399...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <th>11136</th>\n",
       "      <th>12457</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_name=AL627309.3;source_gene=ENSG000002399...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <th>11136</th>\n",
       "      <th>11635</th>\n",
       "      <th>.</th>\n",
       "      <th>-</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_name=AL627309.3;source_gene=ENSG000002399...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>11630</th>\n",
       "      <th>13433</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_name=AP006222.2;source_gene=ENSG000002864...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <th>11630</th>\n",
       "      <th>13433</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>gene_name=AP006222.2;source_gene=ENSG000002864...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">CAT</th>\n",
       "      <th>exon</th>\n",
       "      <th>371998</th>\n",
       "      <th>372122</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>.</th>\n",
       "      <td>source_transcript=ENST00000618181.4;source_tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <th>371998</th>\n",
       "      <th>372122</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>2</th>\n",
       "      <td>source_transcript=ENST00000618181.4;source_tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">exon</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">371998</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">372122</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">.</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">+</th>\n",
       "      <th>.</th>\n",
       "      <td>source_transcript=ENST00000618323.4;source_tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>.</th>\n",
       "      <td>source_transcript=ENST00000618779.4;source_tra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <th>371998</th>\n",
       "      <th>372122</th>\n",
       "      <th>.</th>\n",
       "      <th>+</th>\n",
       "      <th>2</th>\n",
       "      <td>source_transcript=ENST00000618779.4;source_tra...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                               ##gff-version 3\n",
       "chr1 Liftoff gene       11136  12457  . - .  gene_name=AL627309.3;source_gene=ENSG000002399...\n",
       "             transcript 11136  12457  . - .  gene_name=AL627309.3;source_gene=ENSG000002399...\n",
       "             exon       11136  11635  . - .  gene_name=AL627309.3;source_gene=ENSG000002399...\n",
       "             gene       11630  13433  . + .  gene_name=AP006222.2;source_gene=ENSG000002864...\n",
       "             transcript 11630  13433  . + .  gene_name=AP006222.2;source_gene=ENSG000002864...\n",
       "...                                                                                        ...\n",
       "     CAT     exon       371998 372122 . + .  source_transcript=ENST00000618181.4;source_tra...\n",
       "             CDS        371998 372122 . + 2  source_transcript=ENST00000618181.4;source_tra...\n",
       "             exon       371998 372122 . + .  source_transcript=ENST00000618323.4;source_tra...\n",
       "                                          .  source_transcript=ENST00000618779.4;source_tra...\n",
       "             CDS        371998 372122 . + 2  source_transcript=ENST00000618779.4;source_tra...\n",
       "\n",
       "[1000 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chm13t2t_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8 (XPython)",
   "language": "python",
   "name": "xpython"
  },
  "language_info": {
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
